openstreetmap-carto.lua

```diff
@@ -303,134 +594,190 @@ function filter_tags_generic(tags)
+        return
     end
 
-   -- Search through object's tags
-    for k, v in pairs(tags) do
-        -- Check if it has a polygon key and not a linestring override, or a polygon k=v
-        for _, ptag in ipairs(polygon_keys) do
-            if k == ptag and v ~= "no" and not (linestring_values[k] and linestring_values[k][v]) then
-                return 1
-            end
+    local area_tags = isarea(object.tags)
+    if object.is_closed and area_tags then
+        add_polygon(object)
+
+        if z_order(object.tags) ~= nil then
+            add_transport_polygon(object)
         end
+    else
+        add_line(object)
 
-        if (polygon_values[k] and polygon_values[k][v]) then
-            return 1
+        if z_order(object.tags) ~= nil then
+            add_transport_line(object)
+        end
+
+        if roads(object.tags) then
+            add_roads(object)
         end
     end
-    return 0
 end
 
-function is_in (needle, haystack)
-    for index, value in ipairs (haystack) do
-        if value == needle then
-            return true
+function osm2pgsql.process_relation(object)
+    -- grab the type tag before filtering tags
+    local type = object.tags.type
+    object.tags.type = nil
+
+    if clean_tags(object.tags) then
+        return
+    end
+    if type == "boundary" then
+        add_line(object)
+
+        if roads(object.tags) then
+            add_roads(object)
+        end
+
+        add_polygon(object)
+
+    elseif type == "multipolygon" then
+        add_polygon(object)
+
+        if z_order(object.tags) ~= nil then
+            add_transport_polygon(object)
         end
+    elseif type == "route" then
+        add_route(object)
     end
-    return false
 end
 
---- Normalizes layer tags
--- @param v The layer tag value
--- @return An integer for the layer tag
-function layer (v)
-    return v and string.find(v, "^-?%d+$") and tonumber(v) < 100 and tonumber(v) > -100 and v or nil
+function osm2pgsql.select_relation_members(relation)
+    if relation.tags.type == 'boundary'
+       and relation.tags.boundary == 'administrative' then
+        local admin = tonumber(admin_level(relation.tags.admin_level))
+        if admin ~= nil then
+            for _, ref in ipairs(osm2pgsql.way_member_ids(relation)) do
+                -- Store the lowest admin_level, and how many relations it used in
+                if not phase2_admin_ways_level[ref] then
+                    phase2_admin_ways_level[ref] = admin
+                    phase2_admin_ways_parents[ref] = 1
+                else
+                    if phase2_admin_ways_level[ref] == admin then
+                        phase2_admin_ways_parents[ref] = phase2_admin_ways_parents[ref] + 1
+                    elseif admin < phase2_admin_ways_level[ref] then
+                        phase2_admin_ways_level[ref] = admin
+                        phase2_admin_ways_parents[ref] = 1
+                    end
+                end
+            end
```

Comment:
This part (line 766-779) is wrong here. It needs to go into the process_relation() function. select_relation_member() must only compute dependencies. Anything related to collecting information for the ways must happen in proccess_relation().

Reply:
+1 This bug affected me when I was trying to set up this style. When updates were applied, the `admin_level` on a `planet_osm_admins` row would change to a wrong value.

Reply:
I think this also serves as a reminder that with the level of custom data import complexity through lua we have here, proper testing is a serious concern.  If what @amandasaurus indicates is right, that this causes problems only with updates and not with a static import then this means the normal testing done with a test database (be that manual or systematic) will not catch this kind of issue.

Reply:
On the other hand, you have to take into account that all of this flex style processing is essentially "new", so people simply need to familiarize themselves with developing such Lua style files, and learn how to effectively and correctly use the different types of functions available. Unless you're the developer of osm2pgsql, errors in a style file may therefor not be immediately obvious.

This is also a special case: I have extensively tested this flex file with "import-only" type workflows, where the output is fine, but since my I don't not have a post-processing workflow capable of dealing with "updates", I have never bothered familiarizing myself with how to run updates with osm2pgsql, hence I never spotted this issue.

Most other type errors, would have been spotted in an "import-only" type workflow.

Reply:
See: https://github.com/gravitystorm/openstreetmap-carto/pull/4846

I am not able to test this with updates. I only tested this change in an import only scenario on a local derived version of this style that has some customizations, but the output seemed fine on first inspection.